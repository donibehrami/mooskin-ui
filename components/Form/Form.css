@import '../_utils/globals/screens.css';

.container{
    width: 100%;
    box-sizing: border-box;
}

.container > div{
    padding: 10px;
}

.container button{
    margin: 10px;
}

.form{
    display: block;
}

.horizontal{
    width: 100%;
    display: flex;
    flex-direction: row;
}

.vertical{
    display: flex;
    flex-direction: column;
}


@media (--xSmall){
    .horizontal{display: flex; flex-direction: column}
}

@media (--small){
    .horizontal{display: flex; flex-direction: column}
}