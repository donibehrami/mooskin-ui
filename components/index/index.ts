import '../_utils/globals/global.css';

export { default as Alert, AlertCloseButton, AlertDescription, AlertIcon, AlertTitle } from '../Alert/Alert';
export { default as Skeleton, SkeletonCircle, SkeletonText } from '../Skeleton/Skeleton';
export { default as Drawer, DrawerBody, DrawerCloseButton, DrawerContent, DrawerFooter, DrawerHeader, DrawerOverlay } from '../Drawer/Drawer';
export { default as Stack, HStack, VStack } from '../Stack/Stack';
export { default as Box } from '../Box/Box';
export { default as Carousel } from '../Carousel/Carousel';
export { default as Checkbox, CheckboxDescription, CheckboxIcon, CheckboxLabel } from '../Checkbox/Checkbox';
export { default as DatePicker } from '../DatePicker/DatePicker';
export { default as DateRange } from '../DateRange/DateRange';
export { default as DateSelect } from '../DateSelect/DateSelect';
export { default as Grid, Row, Col } from '../Grid/Grid';
export {
    default as Input,
    InputContainer,
    InputDescription,
    InputEmoji,
    InputIcon,
    InputLabel,
    InputOption,
    InputOptionList,
    InputOptionListTitle,
    InputOverlay
} from '../Input/Input';
export { default as List, ListItem, ListItemBody, ListItemEnd, ListItemHead } from '../List/List';
export { default as Modal, ModalBody, ModalCloseButton, ModalContent, ModalFooter, ModalHeader, ModalOverlay } from '../Modal/Modal';
export { default as Pagination, PaginationButton } from '../Pagination/Pagination';
export { default as Radio, RadioDescription, RadioIcon, RadioLabel } from '../Radio/Radio';
export { default as Sidemenu, SidemenuItem } from '../Sidemenu/Sidemenu';
export { default as IconButton } from '../IconButton/IconButton';
export {
    default as Select,
    SelectContainer,
    SelectDescription,
    SelectFilter,
    SelectIcon,
    SelectLabel,
    SelectLoader,
    SelectOption,
    SelectOptionList,
    SelectOverlay,
    SelectPagination,
    SelectPlaceholder
} from '../Select/Select';
export { default as Switch } from '../Switch/Switch';
export { default as TextArea, TextareaDescription, TextareaLabel } from '../TextArea/TextArea';
export { default as Tabs, TabContent, TabHeader, Tab } from '../Tabs/Tabs';
export { default as Tags, TagClose, TagInput, Tag } from '../Tags/Tags';
export { default as Table, TableHeader, TableHeaderItem, TableRow, TableRowItem } from '../Table/Table';

export { default as IInputCallbackData } from '../_utils';
