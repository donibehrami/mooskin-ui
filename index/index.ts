// Context
export { MooskinContext, MooskinContextProvider, useMooskinContext, withMooskinContext } from '../components/Styled/MooskinContextProvider';

// Components
export { ActionsDropdown, ActionsDropdownArrow, ActionsDropdownItem } from '../components/ActionsDropdown/ActionsDropdown';
export { Alert, AlertCloseButton, AlertDescription, AlertIcon, AlertTitle } from '../components/Alert/Alert';
export { Button } from '../components/Button/Button';
export { Anchor } from '../components/Anchor/Anchor';
export { Skeleton, SkeletonCircle, SkeletonText } from '../components/Skeleton/Skeleton';
export {
    Drawer,
    DrawerBody,
    DrawerCloseButton,
    DrawerContent,
    DrawerFooter,
    DrawerHeader,
    DrawerOverlay
} from '../components/Drawer/Drawer';
export { Stack, HStack, VStack } from '../components/Stack/Stack';
export { Box } from '../components/Box/Box';
export { Carousel } from '../components/Carousel/Carousel';
export { Checkbox, CheckboxDescription, CheckboxIcon, CheckboxLabel } from '../components/Checkbox/Checkbox';
export { DateTimePicker } from '../components/DateTimePicker/DateTimePicker';
export { DatePicker } from '../components/DatePicker/DatePicker';
export { DateRange } from '../components/DateRange/DateRange';
export { DateSelect } from '../components/DateSelect/DateSelect';
export { Description } from '../components/Description/Description';
export { Layout } from '../components/Layout/Layout';
export { Selector, SelectorItem } from '../components/Selector/Selector';
export { Label } from '../components/Label/Label';
export {
    Expandable,
    ExpandableItem,
    ExpandableItemButton,
    ExpandableItemContainer,
    ExpandableItemContent,
    ExpandableItemText
} from '../components/Expandable/Expandable';
export { Grid, Row, Col } from '../components/Grid/Grid';
export {
    Input,
    InputContainer,
    InputEmoji,
    InputIcon,
    InputOption,
    InputOptionList,
    InputOptionListTitle,
    InputOverlay
} from '../components/Input/Input';
export { List, ListItem, ListItemBody, ListItemEnd, ListItemHead } from '../components/List/List';
export { Modal, ModalBody, ModalCloseButton, ModalContent, ModalFooter, ModalHeader, ModalOverlay } from '../components/Modal/Modal';
export { Pagination, PaginationButton } from '../components/Pagination/Pagination';
export { Radio, RadioDescription, RadioIcon, RadioLabel } from '../components/Radio/Radio';
export { Sidemenu, SidemenuItem } from '../components/Sidemenu/Sidemenu';
export { IconButton } from '../components/IconButton/IconButton';
export { Steps, StepContent, StepHeader, Step } from '../components/Steps/Steps';
export {
    Select,
    SelectContainer,
    SelectFilter,
    SelectIcon,
    SelectLoader,
    SelectOption,
    SelectOptionList,
    SelectOverlay,
    SelectPagination,
    SelectPlaceholder
} from '../components/Select/Select';
export { Switch, SwitchHandle } from '../components/Switch/Switch';
export { Loader } from '../components/Loader/Loader';
export { NumberLabel } from '../components/NumberLabel/NumberLabel';
export { LoadingBar } from '../components/LoadingBar/LoadingBar';
export { TextArea } from '../components/TextArea/TextArea';
export { TextEditor } from '../components/TextEditor/TextEditor';
export { Tabs, TabContent, TabHeader, Tab } from '../components/Tabs/Tabs';
export { Tags, TagClose, TagInput, Tag } from '../components/Tags/Tags';
export { Table, TableHeader, TableHeaderItem, TableRow, TableRowItem } from '../components/Table/Table';
export { Footer, FooterBody, FooterEnd, FooterHead } from '../components/Footer/Footer';

// Models
export {
    IActionsDropdownArrowComponentProps,
    IActionsDropdownComponentProps,
    IActionsDropdownItemComponentProps
} from '../components/ActionsDropdown/model';
export { IButtonComponentProps } from '../components/Button/model';
export { IBaseAlertComponentProps } from '../components/Alert/model';
export { IAnchorComponentProps } from '../components/Anchor/model';
export {
    IBaseBoxComponentProps,
    IBoxComponentProps,
    IAnchorBoxComponentProps,
    IBoxLabelComponentProps,
    IButtonBoxComponentProps,
    IInputBoxComponentProps,
    ITextAreaBoxComponentProps
} from '../components/Box/model';
export { ICarouselComponentProps } from '../components/Carousel/model';
export { ICheckboxComponentProps, ICheckboxIconComponentProps } from '../components/Checkbox/model';
export {
    IDatePickerCommonProps,
    IDatePickerComponentProps,
    IDatePickerKeyboardComponentProps,
    PickerType as DatePickerType
} from '../components/DatePicker/model';
export {
    IDateTimePickerCommonProps,
    IDateTimePickerComponentProps,
    IDateTimePickerKeyboardComponentProps,
    PickerType as DateTimePickerType
} from '../components/DateTimePicker/model';
export { IDateRangePickerComponentProps, IRangeSelection } from '../components/DateRange/model';
export { IDateSelectComponentProps } from '../components/DateSelect/model';
export { IDescriptionComponentProps } from '../components/Description/model';
export { ILabelComponentProps } from '../components/Label/model';
export { IDrawerComponentProps, IDrawerContentComponentProps, IDrawerOverlayComponentProps } from '../components/Drawer/model';
export { IExpandableCommonComponentProps, IExpandableComponentProps, IExpandableItemComponentProps } from '../components/Expandable/model';
export { IColProps, IGridProps, IRowProps } from '../components/Grid/model';
export { IIconButtonComponentProps } from '../components/IconButton/model';
export {
    IInputComponentProps,
    IInputContainerComponentProps,
    IInputEmojiComponentProps,
    IInputListComponentProps,
    IInputOptionComponentProps
} from '../components/Input/model';
export { ILayoutComponentProps } from '../components/Layout/model';
export { ILoaderComponentProps } from '../components/Loader/model';
export { ILoadingBarComponentProps } from '../components/LoadingBar/model';
export { IModalComponentProps, IModalContentComponentProps, IModalOverlayComponentProps } from '../components/Modal/model';
export { INumberLabelComponentProps } from '../components/NumberLabel/model';
export { IPaginationButtonComponentProps, IPaginationComponentProps } from '../components/Pagination/model';
export { IRadioComponentProps, IRadioIconComponentProps } from '../components/Radio/model';
export {
    ISelectComponentProps,
    ISelectFilterComponentProps,
    ISelectOptionComponentProps,
    ISelectPaginationComponentProps
} from '../components/Select/model';
export { ISelectorComponentProps, ISelectorItemComponentProps } from '../components/Selector/model';
export { ISidemenuComponentProps, ISidemenuItemComponentProps } from '../components/Sidemenu/model';
export {
    ICommonSkeletonComponentProps,
    ISkeletonCircleComponentProps,
    ISkeletonComponentProps,
    ISkeletonTextComponentProps
} from '../components/Skeleton/model';
export { IStackComponentProps } from '../components/Stack/model';
export { IStepCommonComponentProps, IStepComponentProps, IStepHeaderComponentProps, IStepsComponentProps } from '../components/Steps/model';
export { ISwitchComponentProps, ISwitchHandleComponentProps } from '../components/Switch/model';
export { ITabCommonComponentProps, ITabComponentProps } from '../components/Tabs/model';
export { ITagComponentProps, ITagsComponentProps, ITagsInputComponentProps } from '../components/Tags/model';
export { ITextAreaComponentProps } from '../components/TextArea/model';
export { IPersonalizationTag, ITextEditorComponentProps } from '../components/TextEditor/model';

// Global Styles
export { GlobalStyle } from '../components/Styled/GlobalStyles';

export { default as IInputCallbackData } from '../components/_utils';
