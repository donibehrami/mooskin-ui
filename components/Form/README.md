# MooSkin UI - Form Component

The MooSkin Form Component works similar to the normal HTML `<form></form>` element, but with pre-defined styling and attributes.

### Usage

To start using the Form Component first you have to Import it

```
Import {Form, FormGroup} from '@moosend/mooskin';
```
or modular import
```
// JS
import Form, {FormGroup} from '@moosend/mooskin/lib/Form';

// CSS
import '@moosend/mooskin/lib/Form/style.css';
```

And then you can simply start using it by typing

```
<Form>
    ...Whatever...
</Form>
```

For easy use, the components are named similar to normal HTML components but with a capital first letter.
For better content management, there is the `FormGroup` available.

```
Import {Form, FormGroup} from '@moosend/mooskin';

<Form>
    <FormGroup>
        ...whatever...
    </FormGroup>
</Form>
```

Like the `<form/>` element it will accept given attributes and render differently based on the given attributes

The FormGroup can be set to align elements horizontaly, default is vertically. For example, if we want two groups of <Input />, the groups to be aligned horizontaly, and the <Input /> vertically. We do this by passing the `horizontal` prop to the <FormGroup>.

```
<Form>
    <FormGroup horizontaly>
        <FormGroup>
            <Input />
            <Input />
            <Input />
        </FormGroup>
        <FormGroup>
            <Input />
            <Input />
            <Input />
        </FormGroup>
    </FormGroup>
</Form>
```

### Component Usage within the <Form>

Overall data handling within the <Form> Component will be done using the used components callbacks, dataLabel, values and React State. For Component related `dataLabel` and Callbacks visit the component related docs.

### Examples

#### Input

Lets say we're handling data within the <Form> using React State.

```
this.state = { name: '' };

.
.
.

<Form>
    <FormGroup>
        <Input 
            dataLabel="name"
            value={this.state.value}
            onChange={this.setValue}
        />
    </FormGroup>
</Form>

// event and data types are TypeScript related
getValue = (e: React.ChangeEvent<HTMLElement>, data: InputCallbackData) => {
    data.dataLabel && this.setState({[data.dataLabel]: data.value});
}
```

In this case, on every input change to the <Input> Component, the state property of `name` will get updated. `dataLabel` determines which state property to update. The callback data for the <Input> component is similar to `e.target.value`, which can be used aswell.

#### Switch

```
this.state = { done: false };

.
.
.

<Form>
    <FormGroup>
        <Switch
            dataLabel="done"
            on={this.state.done}
            onClick={this.getValue}
        />
    </FormGroup>
</Form>

// event and data types are TypeScript related
getValue = (e: React.MouseEvent<HTMLElement>, data: InputCallbackData) => {
    data.dataLabel && this.setState({[data.dataLabel]: data.value});
}
```

In this case, on every click to the <Switch> Component, the state property of `done` will get updated. `dataLabel` determines which state property to update. The callback data for the <Switch> component is a boolean which determines if the switch is on or off.

#### Switch

Lets say we're handling data within the <Form> using React State.

```
this.state = { contact: [{}, {}, {}] };

.
.
.

<Form>
    <FormGroup>
       <CheckboxGroup onChange={this.getValue} dataLabel="contact" >
            <CheckBox value="email" label="Email" checked={this.state.contact[0].checked}/>
            <CheckBox value="tel" label="Telephone" checked={this.state.contact[1].checked}/>
            <CheckBox value="raven" label="Raven" checked={this.state.contact[2].checked}/>
        </CheckboxGroup>
    </FormGroup>
</Form>

// event and data types are TypeScript related
getValue = (e: React.MouseEvent<HTMLElement>, data: InputCallbackData) => {
    data.dataLabel && this.setState({[data.dataLabel]: data.value});
}
```

In this case, on every click to the <CheckBox> Component, the state property of `contact` will get updated. `dataLabel` determines which state property to update. The callback data for the <CheckboxGroup> component is an array of objects reflecting the state of each CheckBox.

### onSubmit & Buttons

The `onSubmit` prop of the <Form>, is a function passed to the <Form> that will be called after a <Button> on the <Form> is clicked. Keep in mind that the <Button> Component has a default type of `button`, so if you want to pass the onSubmit prop to buttons inside the form, pass the prop type `submit` to the <Button> (<Button type="submit">). 

```
<Button type="submit">
    Submit
</Button>
```

The Form passes callback data aswell, like most Mooskin Components. Form callback data consists of the values of all Mooskin Components within the <Form>. For Example:

```
this.state = {
    name: '',
    done: false,
    start: ''
};

.
.
.

<Form onSubmit={this.onSubmit}>
    <FormGroup>
        <Input 
            dataLabel="name"
            value={this.state.value}
            onChange={this.setValue}
        />
        <Switch
            dataLabel="done"
            on={this.state.done}
            onClick={this.setValue}
        />
        <Select
            dataLabel="start"
            selected={this.state.start}
            onChange={this.setValue}
        >
            <Option value="asap">Asap</Option>
            <Option value="today">Today</Option>
            <Option value="now">Now</Option>
        </Select>
    </FormGroup>
    <Button>Go!</Button>
</Form>

setValue = (e, data) => {
    data.dataLabel && this.setState({[data.dataLabel]: data.value});
}

onSubmit = (e, data) => {
    console.log(data.value);
    // or manipulate the callback data, which is just an object of data within the Form
}
```

So if we fill up the fields, let's say we type on the `<Input>` a string of `Doni`, click on the `<Switch>` to make it `true`, and select the 1st `Option` on the `<Select>`. The resulting callback data will look something like this:

```
{
    name: 'Doni',
    done: true,
    start: 'asap'
}
```
now you're free to play around with the data, make aSync requests, whatever.

### Validation of components

When the form is submitted, validation functions will be called that has been passed to components within the form. Depending on their return value (true or false) the form will be marked as valid or invalid. If only one validation function returns false the Form will not be submitted, therefore not returning the submit object.

<div class="playground-doc">

## Supported attributes for Form 

* `className` - add additional classes to the Form
* `style` - or additional styles
* `onSubmit` - a function which will handle the data within the form

## Supported attributes for FormGroup

* `horizontal` - elements within FormGroup will align horizontally
* `className` - add additional classes to the FormGroup
* `style` - or additional styles

</div>

Allthough these attributes are supported, all of them are optional.

___

#### For more

[MooSkin-UI](https://github.com/moosend/mooskin-ui)
