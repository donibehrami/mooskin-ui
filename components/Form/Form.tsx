import * as React from 'react';

import styles from './Form.css';

import {IInputCallbackData} from '../../components/_utils/types/commonTypes';

import Button, {IButtonProps} from '../Button/Button';
import {ICheckBoxData, ICheckBoxProps} from '../Checkbox/Checkbox';
import {CheckboxGroup, Input, RadioGroup, Select, Switch, TextArea} from '../index/';
import {IRadioData, IRadioProps} from '../Radio/Radio';

export interface IFormProps{

    /** what data is being used, helps whn extracting user input, you know on what field changes are made */
    dataLabel?: string;

    /** Form children */
    children?: any;

    /** Form classname */
    className?: string;

    /** Form styles */
    style?: React.CSSProperties;

    /** callback data on submit */
    onSubmit?: (e: React.FormEvent<HTMLElement>, data: IInputCallbackData) => void;

    // /** Form action */
    // action?: string;

    // /** Form method */
    // method?: string;

    // /** Form enctype */
    // enctype?: string;

    // /** Form name */
    // name?: string;

}

export interface IFormGroupProps{

    /** Form children */
    children?: any;

    /** Form classname */
    className?: string;

    /** Form styles */
    style?: React.CSSProperties;

    /** wether the content should be aligned horizontaly */
    horizontal?: boolean;

}

export default class Form extends React.Component<IFormProps, {}>{

    public static FormGroup: React.StatelessComponent<IFormGroupProps>;

    public render(){

        const {style, className} = this.props;
        const form = this.assignPropsToChildren();

        return(
            <div style={style} className={`form-component ${styles.container}`}>
                <form
                    className={`form ${styles.form} ${className}`}
                    name={name}
                >
                     {form}
                    {/* {this.props.children} */}
                </form>
            </div>
        );
    }

    private onSubmit = (children: any) => {
        return (e: React.MouseEvent<HTMLElement>) => {
            e.preventDefault();
            const data = this.getEssence(children);
            this.props.onSubmit && this.props.onSubmit(e, {value: data, dataLabel: this.props.dataLabel});
        };
    }

    private assignPropsToChildren = () => {
        const formElements: any = [];
        const buttonProps: Partial<IButtonProps> = {
            onClick: this.onSubmit(this.getChildren())
        };
        React.Children.map(this.props.children, (child, index) => {
            const keyProp: Partial<any & {key: number}> = {
                key: index,
            };
            if (React.isValidElement<IButtonProps>(child)){
                if (child.type === Button && child.props.type === 'submit'){
                    formElements.push(
                        React.cloneElement(child, {...keyProp, ...buttonProps})
                    );
                } else {
                    formElements.push(
                        React.cloneElement(child, keyProp)
                    );
                }
            }
        });

        return formElements;

    }

    private getChildren = () => {
        const formChildren: any = [];
        React.Children.map(this.props.children, (child, index) => {
            if (React.isValidElement(child)){
                formChildren.push(child);
            }
        });

        return formChildren;
    }

    private getEssence = (formChildren: any) => {
        let data: any = {};
        data = this.collectEssence(formChildren, data);
        return data;
    }

    private collectEssence = (formChildren: any, data: any) => {
        if (Array.isArray(formChildren)){
            formChildren.map((element: any) => {
                if (element.type === Input && element.props.value !== undefined && element.props.value !== ''){
                    data[element.props.dataLabel] = element.props.value;
                } else if (element.type === TextArea && element.props.value !== undefined){
                    data[element.props.dataLabel] = element.props.value;
                } else if (element.type === Switch && element.props.on !== undefined){
                    data[element.props.dataLabel] = element.props.on;
                } else if (element.type === Select && element.props.selected !== undefined){
                    data[element.props.dataLabel] = element.props.selected;
                } else if (element.type === RadioGroup){
                    const radios: IRadioData[] = [];
                    element.props.children.map((radio: React.ReactElement<IRadioProps>) => {
                        radios.push(radio.props);
                    });
                    data[element.props.dataLabel] = radios;
                } else if (element.type === CheckboxGroup){
                    const checkboxes: ICheckBoxData[] = [];
                    element.props.children.map((checkbox: React.ReactElement<ICheckBoxProps>) => {
                        checkboxes.push(checkbox.props);
                    });
                    data[element.props.dataLabel] = checkboxes;
                } else {
                    this.collectEssence(element, data);
                    // throw new Error('Elements used within the form are not supported');
                }
            });
        }
        if (formChildren.type === FormGroup){
            this.collectEssence(formChildren.props.children, data);
        }
        return data;
    }
}

export const FormGroup: React.StatelessComponent<IFormGroupProps> = (props) => {

    const alignment = !props.horizontal ? styles.vertical : styles.horizontal;

    return (
        <div style={props.style} className={`formgroup-component ${styles.container} ${alignment}`}>
            {props.children}
        </div>
    );
};
