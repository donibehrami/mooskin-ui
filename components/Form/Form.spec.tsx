import * as React from 'react';
import {Button, Input} from '../index';
import Form, {FormGroup} from './Form';

import {render, shallow} from 'enzyme';

describe('Form', () => {

    test('renders Form correctly', () => {
        const func = jest.fn();

        const tree = shallow(
            <Form className="myClass" dataLabel="SomeForm" onSubmit={func} style={{width: '50px'}}/>
        );

        expect(tree).toMatchSnapshot();
    });

    test('renders FormGroup correctly', () => {

        const tree = shallow(
            <FormGroup className="myClass" style={{width: '50px'}} horizontal/>
        );

        expect(tree).toMatchSnapshot();
    });

    test('renders Form with FormGroup correctly', () => {
        const func = jest.fn();

        const tree = render(
            <Form className="myClass" dataLabel="SomeForm" onSubmit={func} style={{width: '50px'}}>
                <FormGroup className="myClass" style={{width: '50px'}} horizontal/>
            </Form>
        );

        expect(tree).toMatchSnapshot();
    });

    test('renders properly with FormGroup and a child', () => {

        const component = shallow(
            <Form className="myClass" dataLabel="plan">
                <FormGroup className="myClass" horizontal >
                    <Input />
                </FormGroup>
            </Form>
        );

        expect(component.find(Form).children.length).toBe(1);
        expect(component.find(FormGroup).children.length).toBe(1);
        expect(component.find(FormGroup).hasClass('myClass')).toBeTruthy;
    });

    test('renders properly with multiple children', () => {

        const component = shallow(
            <Form className="myClass" dataLabel="plan">
                <FormGroup className="myClass" horizontal >
                    <FormGroup>
                        <Input />
                        <Input />
                        <Input />
                    </FormGroup>
                    <FormGroup>
                        <Input />
                        <Input />
                        <Input />
                    </FormGroup>
                </FormGroup>
            </Form>
        );

        expect(component.find(Form).children.length).toBe(1);
        expect(component.find(FormGroup).length).toBe(3);
        expect(component.find(Input).length).toBe(6);
    });

    test('onSubmit callback function called when button is clicked', () => {
        const func = jest.fn();

        const component = shallow(
            <Form onSubmit={func} style={{color: 'blue'}}>
                <FormGroup>
                    <Input />
                    <Input />
                </FormGroup>
                <Button type="submit"/>
            </Form>
        );

        component.find(Button).simulate('click', { preventDefault: () => undefined });
        expect(func).toHaveBeenCalled();
    });

    test('onSubmit callback ignored when button type is not button, the custom button callback is called', () => {
        const func = jest.fn();
        const func2 = jest.fn();

        const component = shallow(
            <Form onSubmit={func} style={{color: 'blue'}}>
                <FormGroup>
                    <Input />
                    <Input />
                </FormGroup>
                <Button type="reset" onClick={func2} />
            </Form>
        );

        component.find(Button).simulate('click');
        expect(func2).toHaveBeenCalled();
    });

});
