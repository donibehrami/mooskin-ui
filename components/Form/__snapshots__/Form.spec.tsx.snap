// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders Form correctly 1`] = `
<div
  className="form-component container"
  style={
    Object {
      "width": "50px",
    }
  }
>
  <form
    className="form form myClass"
    name="nodejs"
  />
</div>
`;

exports[`Form renders Form with FormGroup correctly 1`] = `
<div
  class="form-component container"
  style="width:50px;"
>
  <form
    class="form form myClass"
    name="nodejs"
  >
    <div
      class="formgroup-component container horizontal"
      style="width:50px;"
    />
  </form>
</div>
`;

exports[`Form renders FormGroup correctly 1`] = `
<div
  className="formgroup-component container horizontal"
  style={
    Object {
      "width": "50px",
    }
  }
/>
`;
